File: examples/shared/decide.mjs
/**
 * THE SHARED HALF OF EVERY EXAMPLE — zero dependencies, Node 18+.
 *
 * Every framework example in this directory does the same four things
 * and differs only in how the framework spells "tool":
 *
 *   1. read the door's 402 (the terms an agent is being asked to sign);
 *   2. ask the store's free preflight whether the door is well-formed,
 *      and its free dry run whether a stock client would sign it —
 *      one POST, one probe, both readings from the same bytes;
 *   3. read network, asset, recipient, amount and the named defects;
 *   4. decide, with every reason named beside the evidence for it.
 *
 * The decision is a derivation, not a score: the same inputs always
 * give the same answer, each `because` line names the check, the
 * advisory or the policy field it came from, and `does_not_establish`
 * rides on every answer so a "pay" is never read as "this merchant
 * delivers". Unknown is never a difference: a probe that could not
 * complete, or a fact the reading does not carry, is `cannot_tell`,
 * never a refusal dressed as one.
 *
 * decide.py is the same logic for the Python frameworks. Both are held
 * to examples/fixtures/expected.json by their own test runners, so the
 * two languages cannot drift apart without the build going red.
 */

export const DEFAULT_BASE = "https://scvd.store";

/** The free doors these examples call, by path. Never a paid door. */
export const DOORS = Object.freeze({
  preflight: "/api/preflight/v2",
  before_you_pay: "/api/before-you-pay/v1",
  conformance: "/api/conformance/v1",
  look: "/api/look/v1",
  verifier_mcp: "/mcp/verifier",
});

/** What no reading here can establish. On every answer, by design. */
export const DOES_NOT_ESTABLISH = Object.freeze([
  "whether the service behind the 402 delivers anything after payment — no probe can",
  "whether the door stays up: the reading was one request at one moment",
  "whether the merchant is honest, or which door you should use — this is evidence, not a recommendation",
]);

function decodeBase64Json(value) {
  const text = Buffer.from(value, "base64").toString("utf8");
  return JSON.parse(text);
}

/**
 * Step 1: GET the door yourself and decode its PAYMENT-REQUIRED
 * challenge. This is what your own x402 client sees; the store's
 * probe sees the same bytes, but reading them on your side means the
 * terms you decide on are the terms you were served.
 */
export async function readChallenge(url, { fetch: fetchImpl = fetch } = {}) {
  const response = await fetchImpl(url, { method: "GET", redirect: "manual" });
  const header = response.headers.get("payment-required");
  const out = { url, status: response.status, x402_version: null, accepts: [], parse_error: null };
  if (response.status !== 402) {
    out.parse_error = `expected 402, got ${response.status}`;
    return out;
  }
  if (!header) {
    out.parse_error = "no PAYMENT-REQUIRED header";
    return out;
  }
  try {
    const challenge = decodeBase64Json(header);
    out.x402_version = challenge.x402Version ?? null;
    out.accepts = Array.isArray(challenge.accepts) ? challenge.accepts : [];
  } catch (error) {
    out.parse_error = `PAYMENT-REQUIRED did not decode: ${error.message}`;
  }
  return out;
}

async function post(base, path, body, fetchImpl) {
  const response = await fetchImpl(`${base}${path}`, {
    method: "POST",
    headers: { "content-type": "application/json", accept: "application/json" },
    body: JSON.stringify(body),
  });
  const json = await response.json();
  if (!response.ok) {
    const error = new Error(`${path} answered ${response.status}: ${json.error ?? "no error text"}`);
    error.status = response.status;
    error.body = json;
    throw error;
  }
  return json;
}

/** Step 2: the door's shape (preflight) — one probe, every check named. */
export function preflight(url, { base = DEFAULT_BASE, fetch: fetchImpl = fetch } = {}) {
  return post(base, DOORS.preflight, { url }, fetchImpl);
}

/**
 * Step 2, both halves at once: the dry run answers "will MY client
 * pay this" and carries the preflight report whole as `the_door`, so
 * the two readings can never be quoted against each other.
 */
export function beforeYouPay(url, { base = DEFAULT_BASE, fetch: fetchImpl = fetch, clientProfile } = {}) {
  const body = clientProfile ? { url, client_profile: clientProfile } : { url };
  return post(base, DOORS.before_you_pay, body, fetchImpl);
}

function termsOf(client, accepts) {
  const chosen = client && client.chosen ? client.chosen : null;
  if (chosen) {
    const accept = accepts && accepts[chosen.index] ? accepts[chosen.index] : null;
    return {
      network: chosen.network,
      asset: chosen.asset,
      pay_to: accept && typeof accept.payTo === "string" ? accept.payTo : null,
      amount_atomic: chosen.amount_atomic,
      amount_usd: chosen.amount_usd ?? null,
      signing_window_seconds: chosen.signing_window_seconds ?? null,
      from: "the accept a stock client would select",
    };
  }
  const first = accepts && accepts[0] ? accepts[0] : null;
  if (!first) return null;
  return {
    network: typeof first.network === "string" ? first.network : null,
    asset: typeof first.asset === "string" ? first.asset : null,
    pay_to: typeof first.payTo === "string" ? first.payTo : null,
    amount_atomic: typeof first.amount === "string" ? first.amount : null,
    amount_usd: null,
    signing_window_seconds: null,
    from: "the first accept in the challenge (no client selection available)",
  };
}

/**
 * Steps 3 and 4. Pure: no network, no clock.
 *
 * @param door    the preflight report (reading.the_door, or preflight()'s answer)
 * @param client  the dry run (reading.your_client), or null when you only ran the preflight
 * @param accepts the challenge's accepts as YOU read them (readChallenge), or omit
 * @param policy  { allowed_networks?: string[], allowed_recipients?: string[], max_amount_usd?: number }
 */
export function decide({ door, client = null, accepts = null, policy = {} } = {}) {
  if (!door || typeof door !== "object") throw new TypeError("decide: a preflight report is required");
  const checks = Array.isArray(door.checks) ? door.checks : [];
  const advisories = Array.isArray(door.advisories) ? door.advisories : [];
  const defects = checks.filter((check) => check && check.ok === false).map((check) => check.name);
  const worthKnowing = advisories.map((advisory) => advisory.name);
  const terms = termsOf(client, accepts);
  const because = [];
  const derivedFrom = { preflight_version: door.version ?? null, checks_run: checks.length, advisories_raised: advisories.length };

  const answer = (decision) => ({
    decision,
    because,
    terms,
    defects,
    worth_knowing: worthKnowing,
    does_not_establish: [...DOES_NOT_ESTABLISH],
    derived_from: derivedFrom,
  });

  if (door.verdict === "unreachable") {
    because.push("preflight verdict is unreachable: the probe itself could not complete, which says nothing about the door (unknown is never a defect)");
    return answer("cannot_tell");
  }
  if (door.verdict === "not_ready") {
    because.push(`preflight verdict is not_ready; failed checks: ${defects.join(", ") || "(none named)"}`);
    return answer("do_not_pay");
  }
  if (door.verdict !== "ready") {
    because.push(`preflight verdict is ${String(door.verdict)}, which this decision does not know how to read`);
    return answer("cannot_tell");
  }
  if (client && client.outcome === "would_throw") {
    because.push(`a stock x402 client would refuse on your machine before signing: ${client.throws_with ?? "(no error text)"}`);
    for (const dropped of client.dropped ?? []) {
      because.push(`accept ${dropped.index} (${dropped.network}, ${dropped.asset}) dropped at ${dropped.stage}: ${dropped.why}`);
    }
    return answer("do_not_pay");
  }
  if (client && client.outcome === "cannot_simulate") {
    because.push("the dry run could not walk the challenge's accepts, so which accept your client would sign is unknown");
    return answer("cannot_tell");
  }
  if (worthKnowing.includes("testnet-network")) {
    because.push("advisory testnet-network: the offer quotes a testnet; a mainnet wallet signing it settles nowhere real");
    return answer("do_not_pay");
  }
  if (!terms) {
    because.push("preflight is ready but no accept could be read, so there are no terms to decide on");
    return answer("cannot_tell");
  }
  if (Array.isArray(policy.allowed_networks)) {
    if (!terms.network) {
      because.push("policy allowed_networks is set and the selected accept names no network");
      return answer("cannot_tell");
    }
    if (!policy.allowed_networks.includes(terms.network)) {
      because.push(`policy allowed_networks does not include ${terms.network}`);
      return answer("do_not_pay");
    }
  }
  if (Array.isArray(policy.allowed_recipients)) {
    if (!terms.pay_to) {
      because.push("policy allowed_recipients is set and the recipient could not be read (pass the challenge's accepts to decide)");
      return answer("cannot_tell");
    }
    if (!policy.allowed_recipients.map((r) => r.toLowerCase()).includes(terms.pay_to.toLowerCase())) {
      because.push(`policy allowed_recipients does not include ${terms.pay_to}`);
      return answer("do_not_pay");
    }
  }
  if (typeof policy.max_amount_usd === "number") {
    if (terms.amount_usd === null || terms.amount_usd === undefined) {
      because.push("policy max_amount_usd is set and the amount in USD could not be resolved for this asset");
      return answer("cannot_tell");
    }
    if (terms.amount_usd > policy.max_amount_usd) {
      because.push(`amount ${terms.amount_usd} USD is above policy max_amount_usd ${policy.max_amount_usd}`);
      return answer("do_not_pay");
    }
  }
  because.push(`preflight verdict is ready (${checks.length} checks, ${defects.length} failed)`);
  if (client) because.push(`a stock client would sign accept ${client.chosen ? client.chosen.index : "?"} (${terms.network}, ${terms.amount_atomic} atomic units)`);
  if (worthKnowing.length > 0) because.push(`advisories outside the verdict, worth reading: ${worthKnowing.join(", ")}`);
  return answer("pay");
}

/**
 * The whole walk in one call, for a framework tool that wants one
 * function: GET the door, one POST to the store, decide.
 */
export async function beforeYouPayWalk(url, { base = DEFAULT_BASE, fetch: fetchImpl = fetch, clientProfile, policy = {} } = {}) {
  const challenge = await readChallenge(url, { fetch: fetchImpl });
  const reading = await beforeYouPay(url, { base, fetch: fetchImpl, clientProfile });
  const decision = decide({ door: reading.the_door, client: reading.your_client, accepts: challenge.accepts, policy });
  return { url, challenge, reading, decision };
}

File: examples/shared/decide.py
"""
THE SHARED HALF OF EVERY PYTHON EXAMPLE — standard library only, 3.9+.

The same four steps as decide.mjs, spelled for CrewAI, PydanticAI and
AutoGen: read the door's 402, ask the store's free dry run (which
carries the free preflight whole as ``the_door``), read the terms and
the named defects, decide with every reason beside its evidence.

decide() is pure and mirrors decide.mjs rule for rule; both are held
to examples/fixtures/expected.json by their own runners, so the two
languages cannot drift apart without the build going red. A decision
is a derivation, never a score; unknown is never a difference.
"""

from __future__ import annotations

import base64
import json
import urllib.error
import urllib.request
from typing import Any, Optional

DEFAULT_BASE = "https://scvd.store"

DOORS = {
    "preflight": "/api/preflight/v2",
    "before_you_pay": "/api/before-you-pay/v1",
    "conformance": "/api/conformance/v1",
    "look": "/api/look/v1",
    "verifier_mcp": "/mcp/verifier",
}

DOES_NOT_ESTABLISH = (
    "whether the service behind the 402 delivers anything after payment — no probe can",
    "whether the door stays up: the reading was one request at one moment",
    "whether the merchant is honest, or which door you should use — this is evidence, not a recommendation",
)


class _NoRedirect(urllib.request.HTTPRedirectHandler):
    def redirect_request(self, req, fp, code, msg, headers, newurl):  # noqa: D401
        return None


_OPENER = urllib.request.build_opener(_NoRedirect)


def read_challenge(url: str, opener=None) -> dict:
    """Step 1: GET the door and decode its PAYMENT-REQUIRED challenge."""
    opener = opener or _OPENER
    out = {"url": url, "status": None, "x402_version": None, "accepts": [], "parse_error": None}
    try:
        with opener.open(urllib.request.Request(url, method="GET"), timeout=20) as response:
            status, headers = response.status, response.headers
    except urllib.error.HTTPError as error:
        status, headers = error.code, error.headers
    out["status"] = status
    if status != 402:
        out["parse_error"] = f"expected 402, got {status}"
        return out
    header = headers.get("payment-required")
    if not header:
        out["parse_error"] = "no PAYMENT-REQUIRED header"
        return out
    try:
        challenge = json.loads(base64.b64decode(header).decode("utf-8"))
        out["x402_version"] = challenge.get("x402Version")
        accepts = challenge.get("accepts")
        out["accepts"] = accepts if isinstance(accepts, list) else []
    except (ValueError, TypeError) as error:
        out["parse_error"] = f"PAYMENT-REQUIRED did not decode: {error}"
    return out


def _post(base: str, path: str, body: dict, opener=None) -> dict:
    opener = opener or _OPENER
    request = urllib.request.Request(
        f"{base}{path}",
        data=json.dumps(body).encode("utf-8"),
        headers={"content-type": "application/json", "accept": "application/json"},
        method="POST",
    )
    try:
        with opener.open(request, timeout=30) as response:
            return json.loads(response.read().decode("utf-8"))
    except urllib.error.HTTPError as error:
        text = error.read().decode("utf-8", "replace")
        try:
            detail = json.loads(text).get("error", "no error text")
        except ValueError:
            detail = text[:200]
        raise RuntimeError(f"{path} answered {error.code}: {detail}") from None


def preflight(url: str, base: str = DEFAULT_BASE, opener=None) -> dict:
    """Step 2: the door's shape — one probe, every check named."""
    return _post(base, DOORS["preflight"], {"url": url}, opener)


def before_you_pay(url: str, base: str = DEFAULT_BASE, client_profile: Optional[dict] = None, opener=None) -> dict:
    """Step 2, both halves: will MY client pay this, with the preflight whole as the_door."""
    body = {"url": url, "client_profile": client_profile} if client_profile else {"url": url}
    return _post(base, DOORS["before_you_pay"], body, opener)


def _terms_of(client: Optional[dict], accepts: Optional[list]) -> Optional[dict]:
    chosen = (client or {}).get("chosen")
    if chosen:
        accept = accepts[chosen["index"]] if accepts and len(accepts) > chosen["index"] else None
        pay_to = accept.get("payTo") if isinstance(accept, dict) and isinstance(accept.get("payTo"), str) else None
        return {
            "network": chosen.get("network"),
            "asset": chosen.get("asset"),
            "pay_to": pay_to,
            "amount_atomic": chosen.get("amount_atomic"),
            "amount_usd": chosen.get("amount_usd"),
            "signing_window_seconds": chosen.get("signing_window_seconds"),
            "from": "the accept a stock client would select",
        }
    first = accepts[0] if accepts else None
    if not isinstance(first, dict):
        return None
    s = lambda v: v if isinstance(v, str) else None  # noqa: E731
    return {
        "network": s(first.get("network")),
        "asset": s(first.get("asset")),
        "pay_to": s(first.get("payTo")),
        "amount_atomic": s(first.get("amount")),
        "amount_usd": None,
        "signing_window_seconds": None,
        "from": "the first accept in the challenge (no client selection available)",
    }


def decide(door: dict, client: Optional[dict] = None, accepts: Optional[list] = None, policy: Optional[dict] = None) -> dict:
    """Steps 3 and 4. Pure: no network, no clock. Mirrors decide.mjs rule for rule."""
    if not isinstance(door, dict):
        raise TypeError("decide: a preflight report is required")
    policy = policy or {}
    checks = door.get("checks") if isinstance(door.get("checks"), list) else []
    advisories = door.get("advisories") if isinstance(door.get("advisories"), list) else []
    defects = [c.get("name") for c in checks if isinstance(c, dict) and c.get("ok") is False]
    worth_knowing = [a.get("name") for a in advisories if isinstance(a, dict)]
    terms = _terms_of(client, accepts)
    because: list = []
    derived_from = {"preflight_version": door.get("version"), "checks_run": len(checks), "advisories_raised": len(advisories)}

    def answer(decision: str) -> dict:
        return {
            "decision": decision,
            "because": because,
            "terms": terms,
            "defects": defects,
            "worth_knowing": worth_knowing,
            "does_not_establish": list(DOES_NOT_ESTABLISH),
            "derived_from": derived_from,
        }

    verdict = door.get("verdict")
    if verdict == "unreachable":
        because.append("preflight verdict is unreachable: the probe itself could not complete, which says nothing about the door (unknown is never a defect)")
        return answer("cannot_tell")
    if verdict == "not_ready":
        because.append(f"preflight verdict is not_ready; failed checks: {', '.join(d for d in defects if d) or '(none named)'}")
        return answer("do_not_pay")
    if verdict != "ready":
        because.append(f"preflight verdict is {verdict}, which this decision does not know how to read")
        return answer("cannot_tell")
    if client and client.get("outcome") == "would_throw":
        because.append(f"a stock x402 client would refuse on your machine before signing: {client.get('throws_with') or '(no error text)'}")
        for dropped in client.get("dropped") or []:
            because.append(f"accept {dropped.get('index')} ({dropped.get('network')}, {dropped.get('asset')}) dropped at {dropped.get('stage')}: {dropped.get('why')}")
        return answer("do_not_pay")
    if client and client.get("outcome") == "cannot_simulate":
        because.append("the dry run could not walk the challenge's accepts, so which accept your client would sign is unknown")
        return answer("cannot_tell")
    if "testnet-network" in worth_knowing:
        because.append("advisory testnet-network: the offer quotes a testnet; a mainnet wallet signing it settles nowhere real")
        return answer("do_not_pay")
    if not terms:
        because.append("preflight is ready but no accept could be read, so there are no terms to decide on")
        return answer("cannot_tell")
    allowed_networks = policy.get("allowed_networks")
    if isinstance(allowed_networks, list):
        if not terms.get("network"):
            because.append("policy allowed_networks is set and the selected accept names no network")
            return answer("cannot_tell")
        if terms["network"] not in allowed_networks:
            because.append(f"policy allowed_networks does not include {terms['network']}")
            return answer("do_not_pay")
    allowed_recipients = policy.get("allowed_recipients")
    if isinstance(allowed_recipients, list):
        if not terms.get("pay_to"):
            because.append("policy allowed_recipients is set and the recipient could not be read (pass the challenge's accepts to decide)")
            return answer("cannot_tell")
        if terms["pay_to"].lower() not in [r.lower() for r in allowed_recipients]:
            because.append(f"policy allowed_recipients does not include {terms['pay_to']}")
            return answer("do_not_pay")
    cap = policy.get("max_amount_usd")
    if isinstance(cap, (int, float)) and not isinstance(cap, bool):
        if terms.get("amount_usd") is None:
            because.append("policy max_amount_usd is set and the amount in USD could not be resolved for this asset")
            return answer("cannot_tell")
        if terms["amount_usd"] > cap:
            because.append(f"amount {terms['amount_usd']} USD is above policy max_amount_usd {cap}")
            return answer("do_not_pay")
    because.append(f"preflight verdict is ready ({len(checks)} checks, {len(defects)} failed)")
    if client:
        index = client["chosen"]["index"] if client.get("chosen") else "?"
        because.append(f"a stock client would sign accept {index} ({terms['network']}, {terms['amount_atomic']} atomic units)")
    if worth_knowing:
        because.append(f"advisories outside the verdict, worth reading: {', '.join(w for w in worth_knowing if w)}")
    return answer("pay")


def before_you_pay_walk(url: str, base: str = DEFAULT_BASE, client_profile: Optional[dict] = None, policy: Optional[dict] = None, opener=None) -> dict:
    """The whole walk in one call: GET the door, one POST to the store, decide."""
    challenge = read_challenge(url, opener)
    reading = before_you_pay(url, base, client_profile, opener)
    decision = decide(reading["the_door"], reading.get("your_client"), challenge["accepts"], policy)
    return {"url": url, "challenge": challenge, "reading": reading, "decision": decision}
